chore: add modular models blog and explainer doc (#697)
openfga/openfga.dev: docs/content/modeling/modular-models.mdx

---
sidebar_position: 6
slug: /modeling/modular-models
description: Modular Models
---

import {
  AuthzModelSnippetViewer,
  CardBox,
  CheckRequestViewer,
  DocumentationNotice,
  Playground,
  ProductConcept,
  ProductName,
  ProductNameFormat,
  RelatedSection,
  RelationshipTuplesViewer,
  SyntaxFormat,
  WriteRequestViewer,
  SupportedLanguage,
} from '@components/Docs';

# Modular Models

<DocumentationNotice />

Modular models allows splitting your authorization module across multiple files and modules, improving upon some of the challenges that may be faced when operating an authorization model within a company, such as:

- A model can grow large and difficult to understand
- As more teams begin to contribute to a model, the ownership boundaries may not be clear and code review processes might not scale

> With modular models, a single model can be split across multiple files in a project. Organized in a way that makes sense for the project or teams, and then ownership for reviews can be expressed using a feature such as [GitHub](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) or [GitLab](https://docs.gitlab.com/ee/user/project/codeowners/) code owners.

Reviewer: ```suggestion
With modular models, a single model can be split across multiple files in a project. It can be organized in a way that makes sense for the project or teams collaborating on it, and it enables ownership for reviews to be expressed using a feature such as [GitHub's](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners), [GitLab's](https://docs.gitlab.com/ee/user/project/codeowners/) or [Gitea's](https://docs.gitea.com/usage/code-owners) code owners.
```

## Key Concepts

### `fga.mod`

> The `fga.mod` is the project file for modular models. This file is used to track the schema version for the final combined model and the individual files that make up the modular model.

Reviewer: ```suggestion
The `fga.mod` is the project file for modular models. This file specifies the schema version for the final combined model and lists the individual files that make up the modular model.
```

| Property | Description |
| -------- | -------- | 
| `schema` | The schema version to be used for the combined model |
| `contents` | The individual files that make up the modular model |

### Modules

> Modules are declared using the `module` keyword in the DSL, and a module can be written across multiple files.

Reviewer: ```suggestion
Modules are declared using the `module` keyword in the DSL, and a module can be written across multiple files. A single file cannot have more than one module though.
```

> Currently, modules are stored as metadata but are not used by <ProductName format={ProductNameFormat.ShortForm}/>. However, modules will be the basis of future features such as applying authorization to writing and reading/querying tuples.

Reviewer: ```suggestion
Currently, modules are stored as metadata but are not used by <ProductName format={ProductNameFormat.ShortForm}/>. In the future, module metadata an be used in upcoming features, such as applying authorization to writing and reading/querying tuples.
```

Author: @rhamzeh what do you think of this change to your suggestions?
```suggestion
Currently, modules are stored as metadata but are not used by <ProductName format={ProductNameFormat.ShortForm}/>. Module metadata will be used in upcoming features, such as applying authorization to writing and reading/querying tuples.
```

### Type Extensions

As teams implement features, they might find that core types they are dependent upon might not contain all the realtions they need. However, it might not make sense for these relations to be owned by the owner of that type if they aren't needed across the system.

To solve this, individual types can be extended within other modules to implement the relations needed.

In order to allow this, there are certain requirements for type extension:

- The extended type must exist
- A single type can only be extended once per file
- The relations added must not already exist, or be part of another type extension

## Example

In this example we'll look at how an authorization model for a SaaS compny with a issue tracking and wiki software could implement modular models.

### Core

It's likely there will be a core set of types owned by a team that manages the overall identity for the company, this would provide the basics usch as users, organizations and groups that can be used by each respective product area.

> ```

Reviewer: ```openfga.dsl
```
```

Author: GitHub really doesn't like that suggestion so I'll apply the change locally

Author: Fixed and added new keywords in 81915259ba263b29cc9875d9354804a0a3ff82ff

module core

type user

type organization
  relations
    define member: [user]
    define admin: [user]

type group
  relations
    define member: [user]
```

### Issue tracking

> In the issue tracking software we'd likely separate out the project and issue model into separate files, we'll also extend the `organization` type here so that we can add a relation to track who can create a project.

Reviewer: ```suggestion
In the issue tracking software we'd likely separate out the project and issue related types into separate files, we'll also extend the `organization` type here so that we can add a relation specific to the issue tracking feature, which is the ability to authorize who can create a project.
```

```
module issue-tracker

> extend type organization

Reviewer: ```suggestion
extend type organization
  relations
```

    define viewer: member from organization
```

```
module issue-tracker

type ticket
  relations
    define project: [project]
    define owner: [user]
```

### Wiki

Our wiki model we'll handle in one file for now until it grows some more. Again, we'll also extend the `organization` type here so that we can add a relation to track who can create a space.

```
module wiki

extend type organization
  relations
    define can_create_space: admin

type space
  relations
    define organization: [organization]
    define can_view_pages: member from organization

type page
  relations
    define space: [space]
    define owner: [user]
```

### `fga.mod`

In order to deploy this model we'll need to create our `fga.mod` manifest file, in here we'll set our schema version and list the individual module files that make up our model.

```yaml
schema: '1.2'
contents:
  - core.fga
  - issue-tracker/projects.fga
  - issue-tracker//tickets.fga
  - wiki.fga
```

### Putting it all together

Now that we have our individual parts of the modular model, we're able to write this model to <ProductName format={ProductNameFormat.ShortForm}/> and then run tests against it.

> In order to write out model we need to use the CLI and run

Reviewer 2: Typo (should it be "our model"?). Also missing punctuation:

```suggestion
In order to write our model we need to use the CLI and run:
```

```shell
fga model write --store-id=$FGA_STORE_ID --file fga.mod
```

We can then write tuples and query this model as you would expect with a singular file authorization model.

> 

Reviewer 3: Shouldn't we explain what happens when you retrieve the model with `fga model get`?

Reviewer: I was thinking we do in a follow up so that we can announce today - but am OK waiting till Tuesday morning to announce.

Reviewer 3: If that goes in the documentation page, we can update it whenever we want.

Author: Added a section to the modeling guide that explains the DSL comments as part of 7d25a36be221152d519cdb0ac90ae1f3c0a9feb6
